src/idp2p.ts

```diff
@@ -0,0 +1,163 @@
+import { Doctype } from '@ceramicnetwork/ceramic-common'
+import { DID } from 'dids'
+
+import { Accessors, createAccessors } from './accessors'
+import { IDX_DOCTYPE_CONFIGS, IdxDoctypeName } from './doctypes'
+import { AuthProvider, NewCeramicApi } from './types'
+
+export interface Idp2pOptions {
+  ceramic: NewCeramicApi
+}
+
+export interface AuthenticateOptions {
+  paths?: Array<string>
+  provider?: AuthProvider
+}
+
+export class Idp2p {
+  _ceramic: NewCeramicApi
+  _docQueue: Record<string, Promise<Doctype>> = {}
+  _rootDocId: string | undefined
+
+  accounts: Accessors
+  profiles: Accessors
+
+  constructor({ ceramic }: Idp2pOptions) {
+    this._ceramic = ceramic
+    this.accounts = createAccessors('accounts', this)
+    this.profiles = createAccessors('profiles', this)
+  }
+
+  get ceramic(): NewCeramicApi {
+    return this._ceramic
+  }
+
+  get user(): DID {
+    if (this._ceramic.user == null) {
+      throw new Error('User is not authenticated')
+    }
+    return this._ceramic.user
+  }
+
+  async authenticate(options: AuthenticateOptions = {}): Promise<void> {
+    let provider = this._ceramic.getDIDProvider()
+    if (provider == null) {
+      if (options.provider == null) {
+        throw new Error('Not provider available')
+      } else {
+        // Use provider and attach it to Ceramic instance
+        provider = options.provider
+        this._ceramic.setDIDProvider(provider)
+      }
+    } else if (options.provider != null && options.provider !== provider) {
+      throw new Error('A different provider is already attached to the Ceramic instance')
+    }
+
+    let user = this._ceramic.user
+    if (user == null) {
+      // @ts-ignore: provider type
+      user = new DID(provider)
+      this._ceramic.setUser(user)
+    }
+
+    await user.authenticate()
+  }
+
+  async getUserDocument(): Promise<Doctype> {
```

Comment:
this should use the `did-resolver` package. In case someone is not using 3ID.

Reply:
Should the constructor config accept a `Resolver` instance to use for this, or another way please?
Also, the resolver returns [`DIDDocument`](https://github.com/decentralized-identity/did-resolver/blob/master/src/resolver.ts#L15-L25) so how should we extract the IDX docId from it please?

Reply:
Makes sense that the constructor accepts an optional `Resolver` argument, alternatively resolver configs (so we can make sure to always have the DID resolver for 3ID (from ceramic). 

Good point about the DIDDocument. Ideally we can include the `idx` property in the resolved document. That will likely need a change to the resolver though. Maybe we can use the 3ID specific ceramic document for now and upgrade this later?

Reply:
Yes good idea about the resolver config 👍 
What about the 3ID resolver we should inject anyways, is it provided by 3ID-connect, Ceramic or something else please?

Reply:
We should use the 3ID provider that  is in the js-ceramic monorepo. This is trivial to set up with the `ceramic` instance we have.

Reply:
👍 OK thanks! So should I start using it now even though we wouldn't have access to the `idx` field, or keep using the current logic please?

I think related to that is what should be the behavior in Ceramic when the user authenticates with a DID that is not a 3ID? Could it still add a `idx` field to this DID, or should it create a 3ID anyways and have a provided DID -> 3ID mapping mechanism?

Reply:
Yeah, and you can update the @ceramicnetwork/3id-did-resolver package to expose the idx property. 

Second question is more difficult. We can't really update other DID methods since they have their own ways of updating things. Don't think we need to solve it right now.

Reply:
Yeah, and you can update the @ceramicnetwork/3id-did-resolver package to expose the idx property. 

Second question is more difficult. We can't really update other DID methods since they have their own ways of updating things. Don't think we need to solve it right now.

Reply:
AFAIKT we can't just add the `idx` field but should use a service endpoint, no? cf https://github.com/decentralized-identity/did-resolver/blob/master/src/resolver.ts#L27-L32

If so, should the data be something like this?
```js
{
  "id": "<did>#idx",
  "type": "CeramicIdentityIndex",
  "serviceEndpoint": "ceramic://<docId>"
}
```

Reply:
I'd rather just extend the  did document. Iirc it's even unclear if service endpoints will make it into the core did spec.

Reply:
We can't really rely on that since only `ceramic-core` uses it internally. (correct me if I'm wrong and it's also available in `ceramic-http-client´).

Reply:
I think you're right it seems to be only in core, I'll just keep the same logic here then 👍